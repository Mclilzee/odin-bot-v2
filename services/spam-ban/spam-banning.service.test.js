const SpamBanningService = require("./spam-banning.service");
const config = require("../../config");

function createInteractionMock(message, guild) {
  let replyArg;

  return {
    reply: jest.fn((arg) => {
      replyArg = arg;
    }),

    guild,
    message,

    // Used by service to retrieve the message
    options: { getMessage: () => message },

    getReplyArg: () => replyArg,
    getAuthorSendArg: () => message.getSendArg(),
    getBanArg: () => message.getBanArg(),
    getReactArg: () => message.getReactArg(),
    getChannelSendArg: () =>
      guild.channels.cache.map((c) => c.getSendArg()).find((m) => m),
  };
}

function createMessageMock() {
  let sendArg;
  let banArg;
  let reactArg;

  return {
    author: {
      id: "123",
      send: jest.fn((arg) => {
        sendArg = arg;
      }),
    },
    member: {
      ban: jest.fn((arg) => {
        banArg = arg;
      }),
    },
    react: jest.fn((arg) => {
      reactArg = arg;
    }),
    getSendArg: () => sendArg,
    getBanArg: () => banArg,
    getReactArg: () => reactArg,
  };
}

function createChannelMock(id) {
  let sendArg;
  return {
    id,
    send: jest.fn((arg) => {
      sendArg = arg;
    }),
    getSendArg: () => sendArg,
  };
}

const createGuildMock = () => ({
  channels: {
    cache: [
      createChannelMock("2342314"),
      createChannelMock("101010"),
      createChannelMock("22223333"),
      createChannelMock(config.channels.moderationLog),
      createChannelMock("2302382"),
      createChannelMock("000000"),
    ],
  },
});

describe("Banning spammer with DM enabled", () => {
  let interactionMock;
  beforeEach(() => {
    const messageMock = createMessageMock();
    const guildMock = createGuildMock();
    interactionMock = createInteractionMock(messageMock, guildMock);
  });

  it("Discord ban api is called with the correct reason", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    expect(interactionMock.message.member.ban).toHaveBeenCalledTimes(1);
    expect(interactionMock.getBanArg()).toMatchSnapshot();
  });

  it("Discord message api is called with the correct message", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    expect(interactionMock.message.author.send).toHaveBeenCalledTimes(1);
    expect(interactionMock.getAuthorSendArg()).toMatchSnapshot();
  });

  it("Reacts with the correct emoji", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    expect(interactionMock.message.react).toHaveBeenCalledTimes(1);
    expect(interactionMock.getReactArg()).toMatchSnapshot();
  });

  it("Sends log to the correct channel", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    interactionMock.guild.channels.cache.forEach((channel) => {
      if (channel.id === config.channels.moderationLog) {
        expect(channel.send).toHaveBeenCalledTimes(1);
        expect(interactionMock.getChannelSendArg()).toMatchSnapshot();
      } else {
        expect(channel.send).not.toHaveBeenCalledTimes(1);
      }
    });
  });

  it("Sends back correct interaction reply to calling moderator", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    expect(interactionMock.getReplyArg()).toMatchSnapshot();
  });
});

describe("Banning spammer who has DM set to private", () => {
  let interactionMock;
  beforeEach(() => {
    const guildMock = createGuildMock();
    const messageMock = createMessageMock();
    messageMock.author.send = jest.fn(() => {
      throw new Error("Can't contact user");
    });
    interactionMock = createInteractionMock(messageMock, guildMock);
  });

  it("Discord ban api is called with the correct reason", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    expect(interactionMock.message.member.ban).toHaveBeenCalledTimes(1);
    expect(interactionMock.getBanArg()).toMatchSnapshot();
  });

  it("Discord message api is called and error handled", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    expect(interactionMock.message.author.send).toHaveBeenCalledTimes(1);
  });

  it("Reacts with the correct emoji", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    expect(interactionMock.message.react).toHaveBeenCalledTimes(1);
    expect(interactionMock.getReactArg()).toMatchSnapshot();
  });

  it("Sends log to the correct channel", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    interactionMock.guild.channels.cache.forEach((channel) => {
      if (channel.id === config.channels.moderationLog) {
        expect(channel.send).toHaveBeenCalledTimes(1);
        expect(channel.getSendArg()).toMatchSnapshot();
      } else {
        expect(channel.send).not.toHaveBeenCalledTimes(1);
      }
    });
  });

  it("Sends back correct interaction reply to calling moderator", async () => {
    await SpamBanningService.handleInteraction(interactionMock);
    expect(interactionMock.getReplyArg()).toMatchSnapshot();
  });
});

// describe("Banning spammer that has left the server", () => {
//   let reactArg;
//
//   const messageMock = {
//     author: {
//       id: "123",
//       send: jest.fn(() => {}),
//     },
//     member: null,
//     react: jest.fn((arg) => {
//       reactArg = arg;
//     }),
//   };
//
//   let reply;
//   const interactionMock = {
//     reply: jest.fn((message) => {
//       reply = message;
//     }),
//     options: {
//       getMessage: () => messageMock,
//     },
//     guild: createGuild(),
//   };
//
//   it("Author message sending api is not called", async () => {
//     await SpamBanningService.handleInteraction(interactionMock);
//     expect(messageMock.author.send).not.toHaveBeenCalled();
//   });
//
//   it("Reacts with the correct emoji to automod message", async () => {
//     await SpamBanningService.handleInteraction(interactionMock);
//     expect(messageMock.react).toHaveBeenCalledTimes(1);
//     expect(reactArg).toMatchSnapshot();
//   });
//
//   it("Does not log any channel", async () => {
//     await SpamBanningService.handleInteraction(interactionMock);
//     interactionMock.guild.channels.cache.forEach((channel) => {
//       expect(channel.send).not.toHaveBeenCalled();
//     });
//   });
//
//   it("Sends back correct interaction reply to calling moderator", async () => {
//     await SpamBanningService.handleInteraction(interactionMock);
//     expect(reply).toMatchSnapshot();
//   });
// });
