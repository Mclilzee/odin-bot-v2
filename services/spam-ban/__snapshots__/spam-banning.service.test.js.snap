// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attempting to bann a bot or team member Does not ban admins 1`] = `
{
  "content": "You do not have the permission to ban this user",
  "ephemeral": true,
}
`;

exports[`Attempting to bann a bot or team member Does not ban bots 1`] = `
{
  "content": "You do not have the permission to ban this user",
  "ephemeral": true,
}
`;

exports[`Attempting to bann a bot or team member Does not ban core 1`] = `
{
  "content": "You do not have the permission to ban this user",
  "ephemeral": true,
}
`;

exports[`Attempting to bann a bot or team member Does not ban maintainers 1`] = `
{
  "content": "You do not have the permission to ban this user",
  "ephemeral": true,
}
`;

exports[`Attempting to bann a bot or team member Does not ban moderators 1`] = `
{
  "content": "You do not have the permission to ban this user",
  "ephemeral": true,
}
`;

exports[`Banning spammer that has left the server Reacts with the correct emoji to automod message 1`] = `"❌"`;

exports[`Banning spammer that has left the server Sends back correct interaction reply to calling moderator 1`] = `
{
  "content": "Couldn't bann <@123>. User is not on the server.",
  "ephemeral": true,
}
`;

exports[`Banning spammer who has DM set to private Discord ban api is called with the correct reason 1`] = `
{
  "reason": "Account is compromised",
}
`;

exports[`Banning spammer who has DM set to private Reacts with the correct emoji 1`] = `"✅"`;

exports[`Banning spammer who has DM set to private Sends back correct interaction reply to calling moderator 1`] = `
{
  "content": "Banned <@123> for spam but wasn't able to contact the user.",
  "ephemeral": true,
}
`;

exports[`Banning spammer who has DM set to private Sends log to the correct channel 1`] = `
{
  "embeds": [
    {
      "author": {
        "icon_url": "image.jpg",
        "name": "Ban | bad.spammer",
      },
      "color": 15747399,
      "fields": [
        {
          "inline": true,
          "name": "User",
          "value": "<@123>",
        },
        {
          "inline": true,
          "name": "Moderator",
          "value": "<@007>",
        },
        {
          "inline": true,
          "name": "Reason",
          "value": "Account is compromised and spamming phishing links.",
        },
      ],
      "footer": {
        "text": "ID: 123",
      },
      "timestamp": "2024-01-31T23:00:00.000Z",
    },
  ],
}
`;

exports[`Banning spammer with DM enabled Discord ban api is called with the correct reason 1`] = `
{
  "reason": "Account is compromised",
}
`;

exports[`Banning spammer with DM enabled Discord message api is called with the correct message 1`] = `
{
  "embeds": [
    {
      "description": "Account is compromised and is used to spam phishing links.

  If you would still like to continue using our server, make sure to change your password and recover your account.
  After that send a detailed contact information to appeal the ban on theodinprojectcontact@gmail.com",
      "title": "Banned: Compromised account / Spam",
    },
  ],
}
`;

exports[`Banning spammer with DM enabled Reacts with the correct emoji 1`] = `"✅"`;

exports[`Banning spammer with DM enabled Sends back correct interaction reply to calling moderator 1`] = `
{
  "content": "Banned <@123> for spam successfully.",
  "ephemeral": true,
}
`;

exports[`Banning spammer with DM enabled Sends log to the correct channel 1`] = `
{
  "embeds": [
    {
      "author": {
        "icon_url": "image.jpg",
        "name": "Ban | bad.spammer",
      },
      "color": 15747399,
      "fields": [
        {
          "inline": true,
          "name": "User",
          "value": "<@123>",
        },
        {
          "inline": true,
          "name": "Moderator",
          "value": "<@007>",
        },
        {
          "inline": true,
          "name": "Reason",
          "value": "Account is compromised and spamming phishing links.",
        },
      ],
      "footer": {
        "text": "ID: 123",
      },
      "timestamp": "2024-01-31T23:00:00.000Z",
    },
  ],
}
`;
